File: ios/WILDiOS/BLE/Ce32FrameParser.swift
import Foundation

final class Ce32FrameParser {
    private enum ParseMode {
        case idle
        case framedCommandId
        case framedPayload
        case framedTerminator
        case rawCommandId
        case rawPayload
    }

    private var mode: ParseMode = .idle
    private var commandId: UInt8 = 0
    private var expectedPayloadLength = -1
    private var payload = Data()
    private var outOfFrameBytes = Data()

    private let onOutOfFrameBytes: (Data) -> Void
    private let onFrame: (UInt8, Data) -> Void

    init(
        onOutOfFrameBytes: @escaping (Data) -> Void = { _ in },
        onFrame: @escaping (UInt8, Data) -> Void
    ) {
        self.onOutOfFrameBytes = onOutOfFrameBytes
        self.onFrame = onFrame
        payload.reserveCapacity(512)
        outOfFrameBytes.reserveCapacity(64)
    }

    func push(_ chunk: Data) {
        for byte in chunk {
            consume(byte)
        }
        flushOutOfFrameBytes()
    }

    func reset() {
        resetParserState()
        outOfFrameBytes.removeAll(keepingCapacity: true)
    }

    private func resetParserState() {
        mode = .idle
        commandId = 0
        expectedPayloadLength = -1
        payload.removeAll(keepingCapacity: true)
    }

    private func flushOutOfFrameBytes() {
        guard !outOfFrameBytes.isEmpty else { return }
        onOutOfFrameBytes(outOfFrameBytes)
        outOfFrameBytes.removeAll(keepingCapacity: true)
    }

    private func consume(_ byte: UInt8) {
        switch mode {
        case .idle:
            switch byte {
            case 0x3C:
                flushOutOfFrameBytes()
                startFramedPacket()
            case 0xAD:
                flushOutOfFrameBytes()
                startRawPacket()
            default:
                outOfFrameBytes.append(byte)
            }

        case .framedCommandId:
            if !beginCommand(byte, payloadMode: .framedPayload, zeroPayloadMode: .framedTerminator) {
                restartFromPotentialLeadByte(byte)
            }

        case .framedPayload:
            payload.append(byte)
            if payload.count >= expectedPayloadLength {
                mode = .framedTerminator
            }

        case .framedTerminator:
            if byte == 0x3E {
                emitFrame()
            } else {
                restartFromPotentialLeadByte(byte)
            }

        case .rawCommandId:
            if !beginCommand(byte, payloadMode: .rawPayload, zeroPayloadMode: .rawPayload) {
                restartFromPotentialLeadByte(byte)
            } else if expectedPayloadLength == 0 {
                emitFrame()
            }

        case .rawPayload:
            payload.append(byte)
            if payload.count >= expectedPayloadLength {
                emitFrame()
            }
        }
    }

    private func startFramedPacket() {
        resetParserState()
        mode = .framedCommandId
    }

    private func startRawPacket() {
        resetParserState()
        mode = .rawCommandId
    }

    private func beginCommand(_ value: UInt8, payloadMode: ParseMode, zeroPayloadMode: ParseMode) -> Bool {
        commandId = value
        expectedPayloadLength = Ce32Protocol.payloadLength(for: value) ?? -1
        guard expectedPayloadLength >= 0 else {
            resetParserState()
            return false
        }

        mode = expectedPayloadLength == 0 ? zeroPayloadMode : payloadMode
        return true
    }

    private func emitFrame() {
        let emittedCommandId = commandId
        let emittedPayload = payload
        resetParserState()
        onFrame(emittedCommandId, emittedPayload)
    }

    private func restartFromPotentialLeadByte(_ value: UInt8) {
        resetParserState()
        switch value {
        case 0x3C:
            startFramedPacket()
        case 0xAD:
            startRawPacket()
        default:
            outOfFrameBytes.append(value)
        }
    }
}

File: app/src/main/java/com/wild/android/ble/Ce32FrameParser.kt
package com.wild.android.ble

class Ce32FrameParser(
    private val onOutOfFrameBytes: (ByteArray) -> Unit = {},
    /** Resolves dynamic-length CE64 replies (currently scheduler response 0xD8). */
    private val payloadLengthResolver: (commandId: Int) -> Int? = { null },
    private val onFrame: (commandId: Int, payload: ByteArray) -> Unit,
) {
    private enum class ParseMode {
        Idle,
        FramedCommandId,
        FramedPayload,
        FramedTerminator,
        RawCommandId,
        RawPayload,
    }

    private var mode = ParseMode.Idle
    private var commandId = -1
    private var expectedPayloadLength = -1
    private val payload = ArrayList<Byte>(512)
    private val outOfFrameBytes = ArrayList<Byte>(64)

    fun push(chunk: ByteArray) {
        for (byte in chunk) {
            consume(byte)
        }
        flushOutOfFrameBytes()
    }

    fun reset() {
        resetParserState()
        outOfFrameBytes.clear()
    }

    private fun resetParserState() {
        mode = ParseMode.Idle
        commandId = -1
        expectedPayloadLength = -1
        payload.clear()
    }

    private fun flushOutOfFrameBytes() {
        if (outOfFrameBytes.isEmpty()) {
            return
        }
        onOutOfFrameBytes(outOfFrameBytes.toByteArray())
        outOfFrameBytes.clear()
    }

    private fun consume(byte: Byte) {
        val value = byte.toInt() and 0xFF

        when (mode) {
            ParseMode.Idle -> {
                when (value) {
                    0x3C -> {
                        flushOutOfFrameBytes()
                        startFramedPacket()
                    }

                    // Desktop CE32 transport uses 0xAD as a raw inbound lead byte,
                    // followed by the real command id and a fixed payload without 0x3E.
                    0xAD -> {
                        flushOutOfFrameBytes()
                        startRawPacket()
                    }

                    else -> {
                        outOfFrameBytes += byte
                    }
                }
            }

            ParseMode.FramedCommandId -> {
                if (!beginCommand(value, ParseMode.FramedPayload, ParseMode.FramedTerminator)) {
                    restartFromPotentialLeadByte(value)
                }
            }

            ParseMode.FramedPayload -> {
                if (!appendPayload(byte)) return
                if (payload.size >= expectedPayloadLength) {
                    mode = ParseMode.FramedTerminator
                }
            }

            ParseMode.FramedTerminator -> {
                if (value == 0x3E) {
                    emitFrame()
                } else {
                    restartFromPotentialLeadByte(value)
                }
            }

            ParseMode.RawCommandId -> {
                if (!beginCommand(value, ParseMode.RawPayload, ParseMode.RawPayload)) {
                    restartFromPotentialLeadByte(value)
                } else if (expectedPayloadLength == 0) {
                    emitFrame()
                }
            }

            ParseMode.RawPayload -> {
                if (!appendPayload(byte)) return
                if (payload.size >= expectedPayloadLength) {
                    emitFrame()
                }
            }
        }
    }

    private fun startFramedPacket() {
        resetParserState()
        mode = ParseMode.FramedCommandId
    }

    private fun startRawPacket() {
        resetParserState()
        mode = ParseMode.RawCommandId
    }

    private fun beginCommand(
        value: Int,
        payloadMode: ParseMode,
        zeroPayloadMode: ParseMode,
    ): Boolean {
        commandId = value
        // D8 is self-describing: the first byte echoes D0/D1/.../DB. Never
        // derive its size from the request currently waiting in the host.
        // Late replies can arrive after a timeout or during a different read.
        expectedPayloadLength = if (value == Ce32Protocol.SchedulerResponse) {
            Int.MAX_VALUE
        } else {
            payloadLengthResolver(value) ?: Ce32Protocol.payloadLengthFor(value) ?: -1
        }
        if (expectedPayloadLength < 0) {
            resetParserState()
            return false
        }

        mode = if (expectedPayloadLength == 0) {
            zeroPayloadMode
        } else {
            payloadMode
        }
        return true
    }

    private fun appendPayload(byte: Byte): Boolean {
        if (commandId == Ce32Protocol.SchedulerResponse && payload.isEmpty()) {
            expectedPayloadLength = Ce32Protocol.schedulerResponsePayloadLengthFor(byte.toInt() and 0xFF)
                ?: run { resetParserState(); return false }
        }
        payload += byte
        return true
    }

    private fun emitFrame() {
        val emittedCommandId = commandId
        val emittedPayload = payload.toByteArray()
        resetParserState()
        onFrame(emittedCommandId, emittedPayload)
    }

    private fun restartFromPotentialLeadByte(value: Int) {
        resetParserState()
        when (value) {
            0x3C -> startFramedPacket()
            0xAD -> startRawPacket()
            else -> outOfFrameBytes += value.toByte()
        }
    }
}
